### PR TITLE
bluetooth: controller: mark legacy adv report as discardable

### LINKED ISSUE #70153: driver: bluetooth: userchan: Stuck in rx_thread: calling read()
Hi All, 

When I was testing a Nordic Controller, I discovered that Zephyr's Host HCI might have problems in userchan mode.

Reproduction steps

1. You need to integrate the test patch I provided(https://github.com/LingaoM/zephyr/commit/3cabf9bfe6bb099b6ea5f4fdadc87e5b7d36f761).

2. `west build -b native_sim zephyr/samples/bluetooth/mesh -p auto` (Must be native_sim, so HCI userchan mode is used.)

3. `west build -b bbc_microbit_v2 zephyr/samples/bluetooth/mesh_provisioner -p auto` (This could be the nRF52840 DK either one is fine.) & `west flash`
4. If you don't have a NEW PTS Dongle, please find the nRF52840 Dongle and flash [pts.zip](https://github.com/zephyrproject-rtos/zephyr/files/14586013/pts.zip) (this is the Controller firmware provided by Bluetooth SIG) Use `nrfutil dfu usb-serial -pkg pts.zip  -p /dev/ttyACM0`, or have PTS Dongle(NEW LE Only - up to Core Spec v5.2 supported (Laird Connectivity, part number 451-00004)), direct attach The Controller.
>>>  sudo btattach -B /dev/ttyACM0 -S 1000000 -R
Attaching Primary controller to /dev/ttyACM0
Switched line discipline from 0 to 15
Device index 0 attached
Manufacturer: 2024

6. Run zephyr.exe generated in the second step. Use `sudo build/zephyr/zephyr.exe --bt-dev=hci0`.

7. Open new terminal, `sudo btmon -w btsnoop.pcapng`.
8. Wait patiently for a while, and you can see that the PB-ADV PB-GATT network provision test is being executed repeatedly.
``` C
Waiting for unprovisioned beacon...
Provisioning over PB-GATT dddd0000000000000000000000000000
Waiting for node to be added...
Added node 0x0002
Waiting for unprovisioned beacon...
Provisioning over PB-ADV dddd0000000000000000000000000000
Waiting for node to be added...
Added node 0x0002
Waiting for unprovisioned beacon...
Provisioning over PB-GATT dddd0000000000000000000000000000
```
 9. The problem will arise when the program in step 6 suddenly crashes.
 ``` C
[00:01:32.220,000] <dbg> bt_driver: rx_thread: calling read()
[00:01:32.220,000] <dbg> bt_driver: rx_thread: Calling bt_recv(0x565c5db4)
[00:01:32.220,000] <dbg> bt_hci_core: bt_recv: buf 0x565c5db4 len 59
[00:01:32.220,000] <dbg> bt_driver: rx_thread: calling read()
[00:01:32.220,000] <dbg> bt_driver: rx_thread: Calling bt_recv(0x565c5dd0)
[00:01:32.220,000] <dbg> bt_hci_core: bt_recv: buf 0x565c5dd0 len 47
[00:01:32.220,000] <dbg> bt_driver: rx_thread: calling read()
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 2
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 2
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: bt_gatt_attr_read: handle 0x0000 offset 0 length 5
[00:01:32.220,000] <dbg> bt_gatt: db_hash_gen: Hash: 
                                  fb 77 e2 2d 82 8a 4b 4d  45 c6 c0 32 1e cf 57 af |.w.-..KM E..2..W.
ASSERTION FAIL [err == 0] @ ZEPHYR_BASE/subsys/bluetooth/host/hci_core.c:340
	command opcode 0x2039 timeout with err -11
@ ZEPHYR_BASE/subsys/bluetooth/host/hci_core.c:340
[00:02:32.210,000] <err> os: z_fatal_error: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:02:32.210,000] <err> os: z_fatal_error: Current thread: 0x565caf80 (unknown)
[00:02:32.210,000] <err> os: k_sys_fatal_error_handler: Halting system
Exiting due to fatal error
```
10. You can see from the timestamp that the program suddenly got stuck at `00:01:32.220,000`, but there is available more data in btsnoop.pcapng.

### LINKED ISSUE #74856: Bluetooth: Mesh: discardable pool hardly used if extended advertising enable.
The current mesh application use extended advertising, and extended advertising event not discardable, so the current discarding pool hardly used, but look like occupy many ram footprint.

image

### PR DESCRIPTION
**Pull request by LingaoM**
Mark legacy adv report as discardable, to avoiding dead-lock.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/70153
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/74856

### CODE REVIEW

**Comment by LingaoM** [comment]
Inspired by this PR: https://github.com/zephyrproject-rtos/zephyr/commit/0cf6cac77e274737880de0178b0978aae167e0d9
Although it cannot completely solve the problem, I think it can be significantly reduced, at least better than the current.

So @martintv could you also take a look too, thanks.

**Comment by jhedberg** [comment]
Technically this may be valid, but I don't really like how it blurs the HCI boundary. HCI drivers *should* be responsible solely for HCI transport specific details and ideally not have to peek into the payload of the HCI packets at all. We may want to look at redesigning the HCI layer so that this kind of checks aren't necessary. In the absence of such a more complete solution, maybe an intermediate option is to at least factor out the "discardable" checks into a single place which takes an HCI event payload as input.

File: drivers/bluetooth/hci/h4.c [hunk 1]
```diff
@@ -32,8 +32,22 @@ LOG_MODULE_REGISTER(bt_driver);
...
+		 * We only care about the `num_reports` and `evt_type` fields
+		 * of extended advertising event.
+		 */
+		uint8_t le_ext_adv_hdr[offsetof(struct bt_hci_evt_le_ext_advertising_report,
+				       adv_info[0].addr)];
+	} __packed;
+
+	struct bt_hci_acl_hdr acl;
+	struct bt_hci_iso_hdr iso;
+} __packed;
```
Comment (jhedberg) [inline drivers/bluetooth/hci/h4.c:50]: Why didn't you keep evt/acl/iso in a union? We never process more than one packet at a time, so there shouldn't be any need to keep them in separate places in memory. At the very least something like this should have been properly described and justified in the commit message.

Comment (LingaoM) [reply drivers/bluetooth/hci/h4.c]: Sorry, my brain is short-circuited. .

Comment (LingaoM) [reply drivers/bluetooth/hci/h4.c]: Must be `union` not `struct`.

**Comment by LingaoM** [comment]
> Technically this may be valid, but I don't really like how it blurs the HCI boundary.

I'm actually not satisfied with this PR because not all problems have been solved, such as:
1. If ext_adv->num_reports bigger 1, even event type is BT_HCI_LE_ADV_EVT_TYPE_LEGACY
2. If the air is full of true extended advertising, the work done by this PR will actually not be effective, and there will still be the possibility of dead-lock. 

But considering the real-life scenario, I think the two points mentioned above will not happen in a short time. I think this also gives us a time window to consider how to completely solve the deadlock problem.

**Pull request closed without merging**